RecoBTag/ONNXRuntime/plugins/ParticleTransformerAK4SonicJetTagsProducer.cc

```diff
@@ -0,0 +1,243 @@
+}
+
+void ParticleTransformerAK4SonicJetTagsProducer::acquire(edm::Event const &iEvent, edm::EventSetup const &iSetup, Input &iInput) {
+  edm::Handle<TagInfoCollection> tag_infos;
+  iEvent.getByToken(src_, tag_infos);
+  client_->setBatchSize(tag_infos->size());
+  skippedInference_ = false;
+  if (tag_infos->empty()) return;
+  unsigned int max_n_cpf_counter = 0; 
+  unsigned int max_n_npf_counter = 0; 
+  unsigned int max_n_vtx_counter = 0; 
+
+  // Find the max n_cpf, n_npf and n_vtx among all the jets in an event. 
+  for (unsigned jet_n = 0; jet_n < tag_infos->size(); ++jet_n) {
+    max_n_cpf_counter = std::max(max_n_cpf_counter,
+                         static_cast<unsigned int>(((*tag_infos)[jet_n]).features().c_pf_features.size()));
+    max_n_npf_counter = std::max(max_n_npf_counter,
+                         static_cast<unsigned int>(((*tag_infos)[jet_n]).features().n_pf_features.size()));
+    max_n_vtx_counter = std::max(max_n_vtx_counter,
+                         static_cast<unsigned int>(((*tag_infos)[jet_n]).features().sv_features.size()));
+  }
+  
+  // If an event has no jet, or all jets has zero n_cpf, n_npf and n_vtx, the inference is skipped.
+  if (max_n_cpf_counter == 0 && max_n_npf_counter == 0 && max_n_vtx_counter == 0) {
+    client_->setBatchSize(0);
+    skippedInference_ = true;
+    return;
+  }
+    
+  // all the jets in the same event will fill up the same amount of n_cpf, n_npf, n_vtx and send to server
+  const unsigned int max_n_cpf = std::clamp(max_n_cpf_counter, (unsigned int)0, (unsigned int)25);
+  const unsigned int max_n_npf = std::clamp(max_n_npf_counter, (unsigned int)0, (unsigned int)25);
+  const unsigned int max_n_vtx = std::clamp(max_n_vtx_counter, (unsigned int)0, (unsigned int)5);
+
+  for (unsigned igroup = 0; igroup < input_names_.size(); ++igroup) {
+    const auto &group_name = input_names_[igroup];
+    auto &input = iInput.at(group_name);
+    unsigned target = 0;
+    
+    if (igroup == parT::kChargedCandidates || igroup == parT::kChargedCandidates4Vec) target = std::max((unsigned int)1, max_n_cpf);
+    else if (igroup == parT::kNeutralCandidates || igroup == parT::kNeutralCandidates4Vec) target = std::max((unsigned int)1, max_n_npf);
+    else if (igroup == parT::kVertices || igroup == parT::kVertices4Vec) target = std::max((unsigned int)1, max_n_vtx);
+    
+    input.setShape(0, target);
+    auto tdata = input.allocate<float>(true);
+    for (unsigned jet_n = 0; jet_n < tag_infos->size(); ++jet_n) {
+      const auto &taginfo = (*tag_infos)[jet_n];
+      const auto &features = taginfo.features();
+      auto &vdata = (*tdata)[jet_n];
+      // Loop through the n cpf, and in the case that n_cpf is smaller than max_n_cpf, add padding values to all features
+      if (igroup == parT::kChargedCandidates) {
+        unsigned int n_cpf = features.c_pf_features.size();
+        n_cpf = std::clamp(n_cpf, (unsigned int)0,  (unsigned int)25);
```

Comment:
as above, you can just put `features.c_pf_features.size()` directly inside the `clamp()`

Reply:
also, why is the minimum 0 here, but 1 in the direct inference producer?

Reply:
well I changed the code and remove the "padding" because I don't think we currently have a way to not do padding when sending event info to server... and no need to have clamp 0 and 25 and then clamp on 1 again like in the old code. Now it directly clamp on 1 and 25.